router/core/websocket.go

```diff
@@ -308,6 +311,44 @@ func (h *WebsocketHandler) handleUpgradeRequest(w http.ResponseWriter, r *http.R
 		return
 	}
 
+	// Authenticate the connection using the initial payload
+	fromInitialPayloadConfig := h.config.Authentication.FromInitialPayload
+	if fromInitialPayloadConfig.Enabled {
+		// Setting the initialPayload in the context to be used by the websocketInitialPayloadAuthenticator
+		r = r.WithContext(authentication.WithWebsocketInitialPayloadContextKey(r.Context(), handler.initialPayload))
+
+		// Later check access control after initial payload is read and set into the context
+		validatedReq, err := h.accessController.Access(w, r)
+		if err != nil {
+			statusCode := http.StatusForbidden
+			if errors.Is(err, ErrUnauthorized) {
+				statusCode = http.StatusUnauthorized
+			}
+			http.Error(handler.w, http.StatusText(statusCode), statusCode)
+			handler.writeErrorMessage(requestID, err)
+			handler.Close()
+			return
+		}
+		handler.r = validatedReq
+
+		// Export the token from the initial payload to the request header
+		if fromInitialPayloadConfig.ExportTokenToRequestHeader.Enabled {
+			var initialPayloadMap map[string]interface{}
+			err := json.Unmarshal(handler.initialPayload, &initialPayloadMap)
+			if err != nil {
+				requestLogger.Error("Error parsing initial payload: %v", zap.Error(err))
+				handler.writeErrorMessage(requestID, err)
```

Comment:
should we do `handler.Close()` here?

Reply:
yes, you're right, since the initial payload or the token is invalid I don't see why the web socket should remain open.